File: web/src/domain/asset/depreciation.ts
export type DepreciationMethod = "NONE" | "STRAIGHT_LINE" | "DECLINING_BALANCE"

export interface DepreciationInput {
  baselineValue: number
  baselineDate: string
  salvageValue: number
  annualRateBasisPoints: number
  method: DepreciationMethod
  asOf: string
}

export interface DepreciationResult {
  value: number
  monthsElapsed: number
}

const BASIS_POINTS = 10_000n
const MONTHS_PER_YEAR = 12n

/** Whole months from one date to another, a month counting only once its day of month has passed. */
export function wholeMonthsBetween(from: string, to: string): number {
  const [fy, fm, fd] = from.split("-").map(Number)
  const [ty, tm, td] = to.split("-").map(Number)
  let months = (ty - fy) * 12 + (tm - fm)
  if (td < fd) months -= 1
  return Math.max(0, months)
}

function monthlyStep(value: bigint, rateBasisPoints: number): bigint {
  const divisor = BASIS_POINTS * MONTHS_PER_YEAR
  return (value * BigInt(rateBasisPoints) + divisor / 2n) / divisor
}

/**
 * BR-13: an asset's current value runs from its latest manual valuation (or its purchase) and
 * declines month by month. Whole minor units; the monthly step is rounded half-up so both
 * implementations agree.
 */
export function depreciate(input: DepreciationInput): DepreciationResult {
  const months = wholeMonthsBetween(input.baselineDate, input.asOf)
  const baseline = BigInt(input.baselineValue)
  const floor = BigInt(Math.min(input.salvageValue, input.baselineValue))
  let value = baseline
  if (input.method === "STRAIGHT_LINE") {
    value = baseline - monthlyStep(baseline, input.annualRateBasisPoints) * BigInt(months)
  } else if (input.method === "DECLINING_BALANCE") {
    for (let month = 0; month < months && value > floor; month++) value -= monthlyStep(value, input.annualRateBasisPoints)
  }
  return { value: Number(value > floor ? value : floor), monthsElapsed: months }
}

File: api/src/main/java/iq/mizan/domain/asset/Depreciation.java
package iq.mizan.domain.asset;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

/**
 * BR-13: an asset's current value runs from its latest manual valuation (or its purchase) and
 * declines month by month. Whole minor units; the monthly step is rounded half-up so both
 * implementations agree.
 */
public final class Depreciation {

    private static final long BASIS_POINTS = 10_000;
    private static final long MONTHS_PER_YEAR = 12;

    private Depreciation() {
    }

    public enum Method {
        NONE,
        STRAIGHT_LINE,
        DECLINING_BALANCE
    }

    public record Input(long baselineValue, LocalDate baselineDate, long salvageValue, int annualRateBasisPoints,
                        Method method, LocalDate asOf) {
    }

    public record Result(long value, int monthsElapsed) {
    }

    public static Result compute(Input in) {
        int months = (int) Math.max(0, ChronoUnit.MONTHS.between(in.baselineDate(), in.asOf()));
        long floor = Math.min(in.salvageValue(), in.baselineValue());
        long value = switch (in.method()) {
            case NONE -> in.baselineValue();
            case STRAIGHT_LINE -> in.baselineValue() - monthlyStep(in.baselineValue(), in.annualRateBasisPoints()) * months;
            case DECLINING_BALANCE -> declining(in.baselineValue(), in.annualRateBasisPoints(), months, floor);
        };
        return new Result(Math.max(floor, value), months);
    }

    private static long declining(long baseline, int rateBasisPoints, int months, long floor) {
        long value = baseline;
        for (int month = 0; month < months && value > floor; month++) {
            value -= monthlyStep(value, rateBasisPoints);
        }
        return value;
    }

    private static long monthlyStep(long value, int rateBasisPoints) {
        long divisor = BASIS_POINTS * MONTHS_PER_YEAR;
        return Math.floorDiv(value * rateBasisPoints + divisor / 2, divisor);
    }
}
